### ISSUE TITLE
 TS allows to assign function with non-optional param to function with optional param and does not report error when invoking non-optional function with undefined param

### ISSUE THREAD

**Issue by enterteg** [microsoft/TypeScript#42062]
**TypeScript Version:**  4.1.2

**Search Terms:**
- function property/method declaration errors
- function method arrow error difference

**Expected behavior:**
When we declare MethodSignatre(`(): void`) (for arrow functions declarations it works correctly `() => void`) function with optional parameters and then later we assign to it function with non-optional parameters we get an error that we cannot assign undefined to non-optional params (check playground link for proper TS error for arrow function declarations).

**Actual behavior:**
When we declare MethodSignatre function with optional parameters and we then later assign to it function with non-optional parameters we don't get any errors and we can invoke this function with undefined as param without any errors even function we passed has params declared as non-optional.

**Related Issues:**
Nope.
**Code**
```ts
interface NonOptionalText {
    text: string;
}

interface OptionalText {
    text?: string;
}

interface ManipulateTextMethod {
    manipulateText({ text } : OptionalText): string;
}

interface ManipulateTextArrow {
    manipulateText: ({ text } : OptionalText) => string;
}

function manipulateNonOptionalText({ text }:  NonOptionalText)  {
    return `manipulateTextText: ${text}`;
}

const testMethod: ManipulateTextMethod = {
    manipulateText: manipulateNonOptionalText,
}

const testArrow: ManipulateTextArrow = {
    manipulateText: manipulateNonOptionalText, // with arrow syntax TS catches error
}

testMethod.manipulateText({ text: undefined }) // NOTICE NO TS ERRORS
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function manipulateNonOptionalText({ text }) {
    return `manipulateTextText: ${text}`;
}
const testMethod = {
    manipulateText: manipulateNonOptionalText,
};
const testArrow = {
    manipulateText: manipulateNonOptionalText,
};
testMethod.manipulateText({ text: undefined }); // NOTICE NO TS ERRORS

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgHIHsQHkAOZiZwA2AKhAB5jIDeAUMg8pJQFzIDOYUoA5gNy0AvrVqhIsRClz5CpClTqMm8gPxtO3EPyEix0eEmQBZOCGA4ArkTiQylIxDAALdABMa9RgFtT5qzYg7MAAKamVKZEFkNmkCEGIggEp1Ll4BYVFwfUljX0trW3kAQSgodAB3DyUfM3yAoLZQ8KoomLw4hPlE5ABeAD4OVK10kRgLEAQZEGQavwKIDGx22SCm5ha2NExYla6GRUYoRwsoaYADWbrCygbkABJqdcEzkdoETE5lTgdnNzYTWr+a5gH4udw9KrePJAwLyNiXGGLHbxOSUAA0OjeHyokE4JTK5X+0PmQXxFV6kIYCJJcJmxICSOWKKCaOQAHo2chysBnMg4KVyewAJ7gODkZAkADKyAQNgQTgg7GQ0DKUExuJBjjBADpqfV5GtaeNXBAYKAIO5BN0OWgsCQAJIAYQAoraJdLnQAlT1YT2SoA)

**Comment by MartinJohns** [comment]
[Why are function parameters bivariant?](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant)
